libtcmu.c

```diff
@@ -928,6 +937,52 @@ struct tcmur_handler *tcmu_get_runner_handler(struct tcmu_device *dev)
 	return handler->hm_private;
 }
 
+/* update the ring buffer's tail */
+#define TCMU_UPDATE_RB_TAIL(mb, ent) \
+do { \
+	mb->cmd_tail = (mb->cmd_tail + tcmu_hdr_get_len((ent)->hdr.len_op)) % mb->cmdr_size; \
+} while (0);
+
+static void tcmulib_entry_complete(struct tcmu_device *dev, uint16_t cmd_id,
+				   uint8_t *sense_buf, int result)
+{
+	struct tcmu_mailbox *mb = dev->map;
+	struct tcmu_cmd_entry *ent = (void *) mb + mb->cmdr_off + mb->cmd_tail;
+
+	pthread_spin_lock(&dev->lock);
+	/* current command could be PAD in async case */
+	while (ent != (void *) mb + mb->cmdr_off + mb->cmd_head) {
+		if (tcmu_hdr_get_op(ent->hdr.len_op) == TCMU_OP_CMD)
+			break;
+		TCMU_UPDATE_RB_TAIL(mb, ent);
+		ent = (void *) mb + mb->cmdr_off + mb->cmd_tail;
+	}
+
+	/* cmd_id could be different in async case */
+	if (cmd_id != ent->hdr.cmd_id) {
+		ent->hdr.cmd_id = cmd_id;
+	}
+
+	if (result == TCMU_NOT_HANDLED) {
+		/* Tell the kernel we didn't handle it */
+		char *buf = ent->rsp.sense_buffer;
+
+		ent->rsp.scsi_status = SAM_STAT_CHECK_CONDITION;
+
+		buf[0] = 0x70;	/* fixed, current */
+		buf[2] = 0x5;	/* illegal request */
+		buf[7] = 0xa;
+		buf[12] = 0x20; /* ASC: invalid command operation code */
+		buf[13] = 0x0;	/* ASCQ: (none) */
+	} else if (result == SAM_STAT_CHECK_CONDITION) {
```

Comment:
Here, the SCSI status is only set if it is CHECK CONDITION. This was the reason why the patch initially didn't work in my environment. With the following change, SCSI scan succeeds and sd device properly appears in the system.
> } else {
                if (result == SAM_STAT_CHECK_CONDITION)
                        memcpy(ent->rsp.sense_buffer, sense_buf,
                               TCMU_SENSE_BUFFERSIZE);
                ent->rsp.scsi_status = result;
        }

Reply:
Nice catch. Thanks.